tests/0147-consumer_group_consumer_mock.c

```diff
@@ -0,0 +1,682 @@
+            RD_KAFKA_RESP_ERR_UNKNOWN_MEMBER_ID,
+            RD_KAFKA_RESP_ERR_FENCED_MEMBER_EPOCH};
+        size_t i;
+        for (i = 0; i < RD_ARRAY_SIZE(fenced_errors); i++) {
+                do_test_consumer_group_heartbeat_fenced_error(fenced_errors[i],
+                                                              0);
+                do_test_consumer_group_heartbeat_fenced_error(fenced_errors[i],
+                                                              1);
+        }
+}
+
+/**
+ * @brief Test consumer group behavior with missing topic id when retrieving
+ *        metadata for assigned topics.
+ *        ensuring:
+ *        - initially a partial acknoledgement is started, with an empty list
+ *          (variation 0) or a single topic (variation 1)
+ *        - fetch doesn't start until broker returns an unknown topic id error
+ *        - when error isn't returned anymore the client finishes assigning
+ *          the partition and reads a message.
+ *
+ * @param variation See calling code.
+ */
+static void do_test_metadata_unknown_topic_id_error(int variation) {
+        rd_kafka_mock_cluster_t *mcluster;
+        const char *bootstraps;
+        rd_kafka_topic_partition_list_t *subscription;
+        rd_kafka_t *c;
+        test_timing_t timing;
+        const char *topic  = "do_test_metadata_unknown_topic_id_error";
+        const char *topic2 = "do_test_metadata_unknown_topic_id_error2";
+
+        SUB_TEST_QUICK("variation: %d", variation);
+
+        mcluster = test_mock_cluster_new(1, &bootstraps);
+        rd_kafka_mock_set_group_consumer_heartbeat_interval_ms(mcluster, 500);
+        rd_kafka_mock_topic_create(mcluster, topic, 1, 1);
+        if (variation == 1) {
+                rd_kafka_mock_topic_create(mcluster, topic2, 1, 1);
+        }
+
+        c = create_consumer(bootstraps, topic, rd_false);
+
+        /* Seed the topic with messages */
+        test_produce_msgs_easy_v(topic, 0, 0, 0, 1, 1000, "bootstrap.servers",
+                                 bootstraps, NULL);
+
+        TIMING_START(&timing, "do_test_metadata_unknown_topic_id_error");
+
+        subscription = rd_kafka_topic_partition_list_new(1);
+        rd_kafka_topic_partition_list_add(subscription, topic,
+                                          RD_KAFKA_PARTITION_UA);
+        if (variation == 1) {
+                rd_kafka_topic_partition_list_add(subscription, topic2,
+                                                  RD_KAFKA_PARTITION_UA);
+        }
+
+        rd_kafka_mock_topic_set_error(mcluster, topic,
+                                      RD_KAFKA_RESP_ERR_UNKNOWN_TOPIC_ID);
+
+        TEST_SAY("Subscribing to topic\n");
+        TEST_CALL_ERR__(rd_kafka_subscribe(c, subscription));
+        rd_kafka_topic_partition_list_destroy(subscription);
+
+        TEST_SAY(
+            "Cannot fetch until Metadata calls replies with "
+            "UNKNOWN_TOPIC_ID\n");
+        test_consumer_poll_no_msgs("no messages", c, 0, 1000);
```

Comment:
We should check the assignments as well here as there won't be this topic in the assignments.